# 频变材料的添加  
Djordjevic-Sarkar Model Parameter Calculation
The model parameters can be calculated if the following information about the material is known:

DC conductivity, sDC (optional; default value is 0)  
DC permittivity, eDC (optional; you have to choose the default. See below)  
A triplet of real numbers, consisting of  
the measurement frequency w1 
the real permittivity e1 at this frequency 
the loss tangent tan d1 at this frequency 

![Alt text](image-1.png)

Relative Permittivity:

![Alt text](image-2.png)

Conductivity:

![Alt text](image-3.png)

![Alt text](image-4.png)

![Alt text](image-5.png)

![Alt text](image-6.png)

In [1]:
import sys
sys.path.append(r"C:\work\Study\Script\Ansys\quickAnalyze\FastSim") #添加pyaedt的路径
import pyLayout
pyLayout.log.setLogLevel("INFO")

from pyLayout import Layout
# layout = Layout()
layout = Layout() #least version
# layout.openAedt(r"C:\work\Project\AE\Script\test_pcb\galileo.aedt")
layout.initDesign()

2024/02/01 16:30:31 - INFO: pyLayout Version: V0.6 20231225
2024/02/01 16:30:31 - INFO: Intial aedt desktop Ansoft.ElectronicsDesktop.2024.1
2024/02/01 16:30:31 - INFO: init design: Project75 : EMDesign1
2024/02/01 16:30:31 - INFO: Simulation log recorded in: C:/Users/yguo/OneDrive - ANSYS, Inc/Documents/Ansoft/Project75_EMDesign1.log


## 方式1：按照HFSS UI设定方式生成 DK+ Conductivity

In [2]:
layout.materials.addHFSSDSModle("test1",dk=4.0,df=0.02)

2024/02/01 16:30:31 - INFO: Djordjevic-Sarkar Model Parameter:
e_infi:3.740755252333763 ,e_delta:0.7481510504667526 ,fA:70403.0168767359 ,fB:159154943091.89536 
2024/02/01 16:30:31 - INFO: Djordjevic-Sarkar Model Parameter:
DK:3.740755252333763+0.05113411554862952/2*ln((159154943091.89536**2+freq**2)/(70403.0168767359**2+freq**2))
,Conductivity:1e-12+2*pi*freq*e0*0.05113411554862952*(atan(freq/70403.0168767359)-atan(freq/159154943091.89536))


对比HFSS生成的材料参数和脚本生成的材料参数，两者完全一致：  
Set in HFSS UI:  
DK:3.74076+0.0255671*ln((2.53303e+22+Freq*Freq)/(4.95658e+09+Freq*Freq))  
Cond:1e-12+2.84472e-12*Freq*(atan(Freq/70403)-atan(Freq/1.59155e+11))  
![Alt text](image-7.png)  
![Alt text](image-8.png)  

Set by Script:  
DK:3.740755252333763+0.05113411554862952/2*ln((159154943091.89536**2+freq**2)/(70403.0168767359**2+freq**2)),  
DF:1e-12+2*pi*freq*e0*0.05113411554862952*(atan(freq/70403.0168767359)-atan(freq/159154943091.89536))  
![Alt text](image-12.png)
![Alt text](image-10.png)  

## 方式2 按照Djordjevic-Sarkar Model方式进行输入，频变DK+DF 
![Alt text](image-13.png)  
or  
![Alt text](image-14.png)

可以获取跟HFSS UI设定一致的结果，使用DK/DF形式更加直观  
![Alt text](image-15.png)
![Alt text](image-16.png)

In [3]:
layout.materials.addHFSSDSModle2("test2",dk=4.0,df=0.02)

2024/02/01 16:30:31 - INFO: Causal material set fA as: 70403.0168767359
2024/02/01 16:30:31 - INFO: Djordjevic-Sarkar Model Parameter:
e_infi:3.740755252333763 ,e_delta:0.7481510504667526 ,fA:70403.0168767359 ,fB:159154943091.89536 
2024/02/01 16:30:31 - INFO: Djordjevic-Sarkar Model Parameter:
DK:re(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122102345)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122102345+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))
,DF:-im(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122102345)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122102345+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))/re(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122102345)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122102345+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))


## 方式三，使用Djordjevic-Sarkar Model 的标准形式，
直接输入e_infi,e_delta,fA,fB,cond_dc=1e-12

In [5]:
layout.materials.addStdDSModel("test3",e_infi=3.740755252333763,e_delta=0.7481510504667526,fA=70403.016876735,fB=159154943091.89536 ,cond_dc=1e-12)

2024/02/01 16:31:06 - INFO: Djordjevic-Sarkar Model Parameter:
e_infi:3.740755252333763 ,e_delta:0.7481510504667526 ,fA:70403.016876735 ,fB:159154943091.89536 
2024/02/01 16:31:06 - INFO: Djordjevic-Sarkar Model Parameter:
DK:re(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122101775)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122101775+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))
,DF:-im(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122101775)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122101775+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))/re(3.740755252333763+0.7481510504667526/ln(1000000000000.0001/442355.20122101775)*ln((1000000000000.0001+1j*2*pi*freq)/(442355.20122101775+1j*2*pi*freq))+1e-12/(1j*2*pi*freq*e0))
